Refactor the code that changes state.

hyejineee opened the pull request:
## 과제 목표
상태 관리를 Redux로 변경해도 모든 컴포넌트 테스트가 통과해야 합니다. 추가로 reducer들에 대한 테스트를 추가해 주세요.

## 제한 조건
모든 기능은 TDD로 구현합니다. 컴포넌트와 reducer에 대한 테스트가 존재해야 합니다. 먼저 테스트를 추가하고 commit 한 뒤 실제 기능을 추가하고 commit합니다. commit 메시지 로그는 아래와 같이 남아야 합니다.
> Restaurant에 대한 테스트를 추가한다.
Restaurant을 구현한다.
Restaurant를 리팩토링한다

- if는 사용할 수 있어도 else 사용하실 수 없습니다. GuardClauses 방법을 사용해주세요.
- switch는 사용하실 수 없습니다.
- let은 사용하실 수 없습니다. const만을 사용하여 문제를 해결해주세요.
- 함수 이름과 변수 이름에 줄임말은 사용하실 수 없습니다. 길더라도 명확하게 써주세요.
- indent(인덴트, 들여쓰기) depth를 1로 유지해주세요. 예를 들어, for문 안에 if문이 있으면 indent depth는 2입니다. depth를 줄이는 좋은 방법은 함수(또는 메소드)를 분리하면 됩니다.

## 요구사항

- [x] ESLint를 돌린 결과 아무런 문제가 없어야 합니다.
- [x] 모든 인수 테스트를 통과시켜야 합니다.
- [x] 한 파일에는 하나의 컴포넌트만 있어야 합니다.
- [x] 테스트 커버리지 100%를 달성해야 합니다.

---

hyejineee committed:
Refactor the code that changes state.

hannut91 reviewed (commented):
강의를 보면서 잘 복습하고 계신 것 같아요 ㅎㅎ 따라치는 것보다 보지 않고 스스로 과제를 수행해보는 연습을 해보시면 더 좋을 것 같아요. 더 힘들겠지만 더 많이 배우실 수 있을거에요

hyejineee commented:
@hannut91 넹!! 일단 강의보고 따라하면서 작업하고 revert후에 다시 한번 더 작업하도록 하겠습니다ㅎㅅㅎ 저 이번주는 금,토 쉬는 날이에여!ㅋ큐ㅠㅠ

hyejineee committed:
Add the Redux package

hyejineee committed:
Create the store

hyejineee committed:
Add the useSelector() to the App.jsx

Instead of using the component's state, I added useSelector().

hyejineee committed:
Create a reducer and actions

hyejineee committed:
Create the mock for test

hyejineee committed:
Separate the reducer from store

hyejineee committed:
Add the test for reducer

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,39 @@
+  tasks: [],
+};
+
+export default function reducer(state = initialState, action) {
+  if (action.type === 'updateTaskTitle') {
+    return {
+      ...state,
+      taskTitle: action.payload.taskTitle,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    const { tasks, newId, taskTitle } = state;
+
+    if (!taskTitle) {
+      return state;
+    }
+
+    return {
+      ...state,
+      newId: newId + 1,
+      taskTitle: '',
+      tasks: [...tasks, { id: newId, title: taskTitle }],
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    const { tasks } = state;
+    return {
+      ...state,
+      tasks: tasks.filter((task) => (task.id !== action.payload.id)),
+    };
+  }
+
+  return state;
+}
```
`if`문이나 `switch`문 사용없이 같은 기능을 구현할 수 있을까요?

hyejineee committed:
Add the container component and test

hyejineee committed:
Modify the Item.jsx

hyejineee committed:
Achieve 100% coverage

hyejineee committed:
Revert "Add the Redux package"

This reverts commit 35a7ba8523f36487f26277c4632b5db7a94fa489.

hyejineee committed:
Add the redux and Setting for using redux

hyejineee committed:
Add the test for update task title

hyejineee committed:
Implement the updateTaskTitle function

hyejineee committed:
Refactor the reducer.test.js

hyejineee committed:
Add the test for add task function

hyejineee committed:
Implement the add task function

hyejineee committed:
Add the test for delete task function

hyejineee committed:
Implement delete task function

hyejineee committed:
Add the test for undefined action

hyejineee committed:
Refactor the App.jsx test

hyejineee committed:
Refactor the App.jsx

Reafactor the App.jsx for using redux

hyejineee committed:
Add a test for the ListContainer component

hyejineee committed:
Implement the ListContainer component

Implement the ListContainer component according to ListContainer test.
And Refactor the Page component to use LissContainer.

hyejineee committed:
Add test for InputContainer

hyejineee committed:
Implement InputCountainer component

Implement the InputContainer component according to test.
And Refactor the Page component.

hyejineee committed:
Refactor App component

hyejineee committed:
Refactor overall code

hannut91 commented on src/InputContainer.jsx:
```diff
@@ -0,0 +1,32 @@
+import { updateTaskTitle, addTask } from './actions';
+
+export default function InputContainer() {
+  const { taskTitle } = useSelector((state) => ({
+    taskTitle: state.taskTitle,
+    tasks: state.tasks,
+  }));
```
`taskTitle`밖에 사용하기 있지 않으니 `tasks`는 지워줘도 될 것 같아요

```suggestion
  const { taskTitle } = useSelector((state) => ({
    taskTitle: state.taskTitle,
  }));
```

hyejineee replied:
넵넵!! 피드백에 대한 수정은 과제2 제출하고 나서 해도 될까요??ㅎㅅㅎ

hannut91 replied:
네넵 ㅎㅎ

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,46 @@
+  taskTitle: action.payload.taskTitle,
+});
+
+const actions = {
+  addTask,
+  deleteTask,
+  updateTaskTitle,
+};
```
위의 함수들은 상태를 변경하는 함수라서 reducer함수라고 불러요. 그래서 `reducers`가 좀 더 정확할 것 같아요

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,46 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (Object.keys(actions).includes(action.type)) {
+    return actions[`${action.type}`]({ state, action });
+  }
```
객체의 property가 있으면 redcuer함수를 실행할려고 하신 것 같아요. 

```js
if (!actions[action.type]) {}
```

간단하게 위와 같이 표현할 수 있을 것 같아요

그리고 
```js
actions[action.type]({ state, action });
```

이렇게만 써줘도 될 것 같아요. action.type이 문자열이기 때문에요

hannut91 reviewed (approved):
사소한 수정말고는 더 이상 수정할게 없어보여요 ㅎㅎ 고생하셨어요
